### PR TITLE
Ahf merger tree tool

### PR DESCRIPTION
**Pull request by TobiBu**
some piece of code to read in the merger trees created by AHF's merger tree tool similar to the consistent trees tool.

### CODE REVIEW

**Comment by TobiBu** [comment]
finally I managed to get a version which passes the tests.
I included further a test for the AHF importer.

Let me know your comments...

File: tangos/tools/ahf_merger_tree_importer.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from ..core import get_or_create_dictionary_item
+from ..core.halo_data import HaloLink, HaloProperty
+from .. import config
+from . import GenericTangosTool
+from six.moves import xrange
+import re
+
+class AHFTreeImporter(GenericTangosTool):
+    tool_name = 'import-ahf-trees'
+    tool_description = 'Import properties that were calculated by the halo finder'
```
Comment (apontzen) [inline tangos/tools/ahf_merger_tree_importer.py]: Is this the correct help text? Presumably it's importing the merger tree in fact.

File: tests/test_ahf_trees.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    Mvir_test, = tangos.get_timestep("test_ahf_merger_tree/tiny_000832").calculate_all("Mvir")
+    npt.assert_allclose(Mvir_test, [2.34068e+11,   4.94677e+10,   4.58779e+10,   4.24798e+10,   2.31297e+10,   2.1545e+10,   1.85704e+10,   1.62538e+10,   1.28763e+10])
+
+def test_ahf_merger_tree_import():
+    importer = tools.ahf_merger_tree_importer.AHFTreeImporter()
+    importer.parse_command_line("--for test_ahf_merger_tree".split())
+    with log.LogCapturer():
+        importer.run_calculation_loop()
+        print(tangos.get_halo("test_ahf_merger_tree/tiny.000832/halo_1").all_links)
```
Comment (apontzen) [inline tests/test_ahf_trees.py]: Suggest removing residual print statement

File: tangos/tools/ahf_merger_tree_importer.py (hunk 1 above)
Comment (TobiBu) [reply tangos/tools/ahf_merger_tree_importer.py]: ups, indeed, I will change that text

File: tests/test_ahf_trees.py (hunk 2 above)
Comment (TobiBu) [reply tests/test_ahf_trees.py]: will be done in the next commit, that's a left over from me figuring out how the automated tests work :-)

**Comment by TobiBu** [comment]
Actually, I could create a tutorial for the Gadget plus AHF files since I was struggling a bit making it work. I will have a look at the files in the next week and get back to you once I have a decent draft of the tutorial.

**Comment by apontzen** [comment]
Brilliant, thanks!

**Comment by TobiBu** [comment]
Hi, Andrew,

i finished the tutorial.
the AHF files can be found here: 
https://datashare.mpcdf.mpg.de/s/pedRAah4uZ4Apis

Let me know if you encounter any issues.

File: docs/first_steps.md [hunk 3]
```diff
@@ -5,6 +5,7 @@ There are currently various sample simulations available to illustrate the use o
 
 * A uniform resolution volume run with GADGET and with SubFind halo catalogues - [click here](first_steps_gadget+subfind.md);
 * An alternative set of Rockstar halo catalogues based on the same GADGET simulation as above - [click here](first_steps_gadget+rockstar.md);
+* An alternative set of Rockstar halo catalogues based on the same GADGET simulation as above - [click here](first_steps_gadget+AHF.md);
```
Comment (apontzen) [inline docs/first_steps.md]: I assume this should say AHF halo catalogues...

**Comment by apontzen** [comment]
When I follow the tutorial, I seem to get NDM=0 for all the halos. Is this just me or is there something to be fixed here?

Also the `finder_id`s are very large numbers; perhaps that's correct but just wanted to check.

File: docs/first_steps.md (hunk 3 above)
Comment (TobiBu) [reply docs/first_steps.md]: Ups, stupid copy and paste mistake... I will fix it.

**Comment by TobiBu** [comment]
> When I follow the tutorial, I seem to get NDM=0 for all the halos. Is this just me or is there something to be fixed here?

Indeed, I only get numbers for Ngas...
I can have a look why that is. But for now I don't have any clue where that comes from.
> 
> Also the `finder_id`s are very large numbers; perhaps that's correct but just wanted to check.
Large finder numbers are correct. That's something inherent to AHF plus Gadget.
Alexander Knebe ade that on purpose but I am not entirely sure why.

**Comment by TobiBu** [comment]
I think I made a mistake when creating the AHF files. 
This simulation is a uniform volume sim of pure dark matter, right?
I will ask Alexander Knebe for help. He already assisted when trying to create the merger trees.

**Comment by apontzen** [comment]
Yes, it's a pure DM uniform volume simulation. Sounds like AHF has just misidentified the particle types. I think this is worth resolving, then it looks like the PR will be done :-)

**Comment by TobiBu** [comment]
> Yes, it's a pure DM uniform volume simulation. Sounds like AHF has just misidentified the particle types. I think this is worth resolving, then it looks like the PR will be done :-)

yes, I am on it. Just sent a mail to Alexander Knebe, to ask for help.

**Comment by TobiBu** [comment]
> > Yes, it's a pure DM uniform volume simulation. Sounds like AHF has just misidentified the particle types. I think this is worth resolving, then it looks like the PR will be done :-)
> 
> yes, I am on it. Just sent a mail to Alexander Knebe, to ask for help.

**Comment by TobiBu** [comment]
> Yes, it's a pure DM uniform volume simulation. Sounds like AHF has just misidentified the particle types. I think this is worth resolving, then it looks like the PR will be done :-)

Hi Andrew,

I talked to Alexander, it seems one has to run AHF in DMO mode but then there is no easy way to make AHF spit out the n_gas column when dealing with DMO Gadget sims. That way, the loading of snaps via the stat files fails because pynbody can't find the n_gas column because it does not exist.
An easy fix would be to modify the AHF stat file to allow for DMO input where no n_gas and n_star column exist. Does that sound like a good idea?

cheers
Tobias

**Comment by apontzen** [comment]
I'm not totally following - the problem seemed to be not missing columns, but rather that there _is_ a `n_gas` column which contains the wrong info?

**Comment by apontzen** [comment]
Rereading, I think I see what you mean - is it that you tried without a `n_gas` column but it failed? If so, yes I agree with your strategy. I'd recommend creating a `DefaultValue` class in `translations.py` to return zero by default, then applying that within the `AHFStatFile` class.

**Comment by TobiBu** [comment]
> Rereading, I think I see what you mean - is it that you tried without a `n_gas` column but it failed? If so, yes I agree with your strategy. I'd recommend creating a `DefaultValue` class in `translations.py` to return zero by default, then applying that within the `AHFStatFile` class.

Yeah, sorry for the confusion. The original problem was that I compiled AHF for a hydro run and applying it to a DMO sim which lead to the `n_gas` column containing the same number of particles as the `npart` column.
When compiling AHF for Gadget DMO runs, the output does not contain the column `n_gas` anymore. Which makes sense since there are only dark matter particles. 
I will create a `DefaultValue` class in `translations.py` to fix that issue with the `AHFStatFile`.

**Comment by TobiBu** [comment]
Hi Andrew,

I took a look into `translations.py` and into the class `HaloStatFile` in `__init__.py`
The problem occurs in line 64 in the function `iter_rows_raw`  when the header of the stat file is read and gives a `ValueError: 'n_gas' is not in list` error.

I don't think a `DefaultValue` class would help for this. 
I modified the list comprehension to test if the expected column exist in the file. 
`ids = [0] + [header.index(a) for a in args if a in header else None]`

Unfortunately I had to set the values of that column manually to 0 in the function `_get_values_for_columns`. Setting it to np.nan or None and then apply a `DefaultValue`transformation did not work for me. Hints how to improve the implementation are very welcome.

cheers
Tobias

**Comment by apontzen** [comment]
I put a suggestion in #102 - does this help?

**Comment by TobiBu** [comment]
hi,
did the rebase from master.

**Pull request merged**